## Table of Contents

* [1. Exploratory Data Analysis](#eda)
    * [1.1. Opening the base dataset](#base)
    * [1.2. Data Evaluation](#evaluation)
        * [1.2.1. Selecting the country: Italy](#country)

* [2. Hypothesis Testing](#hypothesis)



<a id="eda"></a>

# 1. EDA

In [70]:
# import opendatasets as od
# import os
import pandas as pd
import numpy as np
import copy
import itertools


<a id="base"></a>

## 1.1. Opening the base dataset

In [18]:
base_rest_df = pd.read_csv('../restaurant_base.csv', sep=';')


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/699545002.py:1: DtypeWarning: Columns (14,16,18,19,20,21,22,23,24,25,26,27,28) have mixed types. Specify dtype option on import or set low_memory=False.
  base_rest_df = pd.read_csv('../restaurant_base.csv', sep=';')


In [19]:
base_rest_df.shape

(198446, 40)

In [20]:
base_rest_df.head()

restaurant_link country                region      province  \
0  g10001637-d10002227  France    Nouvelle-Aquitaine  Haute-Vienne   
1   g10002858-d4586832  France   Centre-Val de Loire         Berry   
2   g10002986-d3510044  France    Nouvelle-Aquitaine       Correze   
3   g10022428-d9767191  France             Occitanie       Aveyron   
4  g10055971-d10040330  France  Auvergne-Rhone-Alpes           Ain   

                       city   latitude  longitude price_level  \
0             Saint-Jouvent  45.961674   1.169131           €   
1                Rivarennes  46.635895   1.386133           €   
2                   Lacelle  45.642610   1.824460           €   
3  Saint-Laurent-de-Levezou  44.208860   2.960470      €€-€€€   
4                  Journans  46.144340   5.331200      €€-€€€   

           cuisines                                           features  ...  \
0            French  Reservations, Seating, Wheelchair Accessible, ...  ...   
1  French, European  Reservations, Seating, Table Service, Wheelcha...  ...   
2            French  Reservations, Seating, Serves Alcohol, Table S...  ...   
3            French  Reservations, Seating, Wheelchair Accessible, ...  ...   
4            French  Reservations, Seating, Table Service, Serves A...  ...   

   meals_Lunch  meals_Breakfast  meals_Drinks  meals_Brunch meals_After-hours  \
0            1                0             0             0                 0   
1            1                0             1             0                 0   
2            1                0             0             0                 0   
3            1                0             0             0                 0   
4            1                0             0             0                 0   

   special_diets_Vegetarian Friendly special_diets_Gluten Free Options  \
0                                  0                                 0   
1                                  0                                 0   
2                                  0                                 0   
3                                  0                                 0   
4                                  0                                 0   

   special_diets_Vegan Options special_diets_Halal special_diets_Kosher  
0                            0                   0                    0  
1                            0                   0                    0  
2                            0                   0                    0  
3                            0                   0                    0  
4                            0                   0                    0  

[5 rows x 40 columns]

In [21]:
base_rest_df.columns

Index(['restaurant_link', 'country', 'region', 'province', 'city', 'latitude',
       'longitude', 'price_level', 'cuisines', 'features', 'avg_rating',
       'total_reviews_count', 'price_min', 'price_max', 'min_open_dttm',
       'min_open_sec', 'max_close_dttm', 'max_close_sec', 'break_flg',
       'max_break_hours', 'max_no_weekend_hours', 'max_weekend_hours',
       'open_mon_hours', 'open_tue_hours', 'open_wed_hours', 'open_thu_hours',
       'open_fri_hours', 'open_sat_hours', 'open_sun_hours', 'meals_Dinner',
       'meals_Lunch', 'meals_Breakfast', 'meals_Drinks', 'meals_Brunch',
       'meals_After-hours', 'special_diets_Vegetarian Friendly',
       'special_diets_Gluten Free Options', 'special_diets_Vegan Options',
       'special_diets_Halal', 'special_diets_Kosher'],
      dtype='object')

<a id="evaluation"></a>

## 1.2. Data Evaluation

In [29]:
drop_columns = [
#     'restaurant_link', 
#     'country', 
    'region', 'province', 'city', 
    'latitude', 'longitude', 
    
#     'price_level', 'cuisines', 'features', 
#     'avg_rating', 'total_reviews_count', 
    'price_min', 'price_max', 
    
    'min_open_dttm',
       'min_open_sec', 'max_close_dttm', 'max_close_sec', 'break_flg',
       'max_break_hours', 'max_no_weekend_hours', 'max_weekend_hours',
       'open_mon_hours', 'open_tue_hours', 'open_wed_hours', 'open_thu_hours',
       'open_fri_hours', 'open_sat_hours', 'open_sun_hours', 
    
#     'meals_Dinner',
#        'meals_Lunch', 'meals_Breakfast', 'meals_Drinks', 'meals_Brunch',
#        'meals_After-hours', 'special_diets_Vegetarian Friendly',
#        'special_diets_Gluten Free Options', 'special_diets_Vegan Options',
#        'special_diets_Halal', 'special_diets_Kosher'
]

In [37]:
base_rest_df2 = base_rest_df.drop(drop_columns, axis=1)
print(base_rest_df2.shape)

(198446, 18)


In [147]:
base_rest_df2.head()

restaurant_link country price_level          cuisines  \
0  g10001637-d10002227  France           €            French   
1   g10002858-d4586832  France           €  French, European   
2   g10002986-d3510044  France           €            French   
3   g10022428-d9767191  France      €€-€€€            French   
4  g10055971-d10040330  France      €€-€€€            French   

                                            features  avg_rating  \
0  Reservations, Seating, Wheelchair Accessible, ...         4.0   
1  Reservations, Seating, Table Service, Wheelcha...         5.0   
2  Reservations, Seating, Serves Alcohol, Table S...         4.0   
3  Reservations, Seating, Wheelchair Accessible, ...         4.5   
4  Reservations, Seating, Table Service, Serves A...         4.0   

   total_reviews_count  meals_Dinner  meals_Lunch  meals_Breakfast  ...  \
0                 36.0             1            1                0  ...   
1                 13.0             1            1                0  ...   
2                 34.0             1            1                0  ...   
3                 11.0             1            1                0  ...   
4                 36.0             1            1                0  ...   

   cuisines_Mediterranean  cuisines_Bar  cuisines_Spanish  cuisines_Cafe  \
0                       0             0                 0              0   
1                       0             0                 0              0   
2                       0             0                 0              0   
3                       0             0                 0              0   
4                       0             0                 0              0   

   cuisines_Pub  cuisines_Seafood  cuisines_Pizza  cuisines_British  \
0             0                 0               0                 0   
1             0                 0               0                 0   
2             0                 0               0                 0   
3             0                 0               0                 0   
4             0                 0               0                 0   

   cuisines_Fast food  cuisines_Asian  
0                   0               0  
1                   0               0  
2                   0               0  
3                   0               0  
4                   0               0  

[5 rows x 31 columns]

In [150]:
# base_rest_df2 = base_rest_df2.drop(['features_cnt'], axis=1)

In [230]:
base_rest_df2['cuisines_cnt'] = base_rest_df2['cuisines'].apply(lambda x: 1+x.count(','))
base_rest_df2['meals_cnt'] = len([col for col in base_rest_df2.columns if 'meals_' in col])
base_rest_df2['special_diets_cnt'] = len([col for col in base_rest_df2.columns if 'special_diets_' in col])

base_rest_df2['price_level_Cheap Eats'] = base_rest_df2['price_level'].apply(lambda x: 1 if str(x) == '€' else 0)
base_rest_df2['price_level_cnt'] = 1

base_rest_df2

restaurant_link  country price_level  \
0       g10001637-d10002227   France           €   
1        g10002858-d4586832   France           €   
2        g10002986-d3510044   France           €   
3        g10022428-d9767191   France      €€-€€€   
4       g10055971-d10040330   France      €€-€€€   
...                     ...      ...         ...   
198441     g946541-d6957229  Romania      €€-€€€   
198442     g946541-d8075038  Romania           €   
198443     g946543-d4308046  Romania      €€-€€€   
198444     g946544-d8490226  Romania      €€-€€€   
198445     g946544-d8749966  Romania      €€-€€€   

                                                 cuisines  \
0                                                  French   
1                                        French, European   
2                                                  French   
3                                                  French   
4                                                  French   
...                                                   ...   
198441         Seafood, European, Mediterranean, Romanian   
198442  International, European, Romanian, Central Eur...   
198443                                     Pizza, Italian   
198444                                Fast food, European   
198445                         Eastern European, Romanian   

                                                 features  avg_rating  \
0       Reservations, Seating, Wheelchair Accessible, ...         4.0   
1       Reservations, Seating, Table Service, Wheelcha...         5.0   
2       Reservations, Seating, Serves Alcohol, Table S...         4.0   
3       Reservations, Seating, Wheelchair Accessible, ...         4.5   
4       Reservations, Seating, Table Service, Serves A...         4.0   
...                                                   ...         ...   
198441  Reservations, Outdoor Seating, Seating, Serves...         4.0   
198442  Reservations, Takeout, Television, Full Bar, A...         3.5   
198443  Delivery, Takeout, Outdoor Seating, Seating, W...         4.0   
198444                                            Seating         3.5   
198445  Reservations, Wheelchair Accessible, Outdoor S...         4.0   

        total_reviews_count  meals_Dinner  meals_Lunch  meals_Breakfast  ...  \
0                      36.0             1            1                0  ...   
1                      13.0             1            1                0  ...   
2                      34.0             1            1                0  ...   
3                      11.0             1            1                0  ...   
4                      36.0             1            1                0  ...   
...                     ...           ...          ...              ...  ...   
198441                 55.0             1            1                0  ...   
198442                 28.0             1            1                1  ...   
198443                 26.0             1            1                0  ...   
198444                 21.0             0            1                0  ...   
198445                 22.0             1            1                0  ...   

        cuisines_Seafood  cuisines_Pizza  cuisines_British  \
0                      0               0                 0   
1                      0               0                 0   
2                      0               0                 0   
3                      0               0                 0   
4                      0               0                 0   
...                  ...             ...               ...   
198441                 1               0                 0   
198442                 0               0                 0   
198443                 0               1                 0   
198444                 0               0                 0   
198445                 0               0                 0   

        cuisines_Fast food  cuisines_Asian  cuisines_cnt  meals_cnt  \


In [31]:
base_rest_df2.groupby('country').size().sort_values(ascending=False)

country
France             44523
Italy              38738
United Kingdom     35535
Spain              30631
Germany            10259
Portugal            8318
Belgium             5824
Greece              5073
The Netherlands     3499
Ireland             2198
Czech Republic      2014
Austria             1995
Poland              1857
Croatia             1679
Sweden              1330
Denmark             1264
Romania             1123
Hungary              885
Finland              648
Bulgaria             536
Slovakia             517
dtype: int64

In [131]:
# rating of values in the column

def rating_of_column_values(df, column, separator=', '):
    df2 = df[list(df)]
    df2['column_2'] = df2[column].str.split(separator)
    tags_list = list(set().union(*df2['column_2'].dropna()))

    tags_list_2 = []
    for tag in tags_list:
        tags_list_2.append([tag, df2['column_2'].dropna().map(set([tag]).issubset).fillna(False).value_counts()[True]])

    tags_list_2 = sorted(tags_list_2, key=lambda x: x[1], reverse=True)
    df2 = df2.drop(['column_2'], axis=1)
    return tags_list_2

# dummies without not popular options of feature columns

def features_to_flags(df, feature_col_list, threshold):
    for col in feature_col_list:
        top_tags = [f[0] for f in rating_of_column_values(df, col, separator=', ') if f[1] >= threshold * len(df)]
        for t in top_tags:
            df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
        df = df.drop(col, axis=1)
    return df


<a id="country"></a>

### 1.2.1. Selecting the country: Italy

In [45]:
italy_df = base_rest_df2[base_rest_df2['country'] == 'Italy']
italy_df2 = features_to_flags(italy_df, ['cuisines', 'features'], 0.05)

italy_df2.columns

/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/334375949.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/334375949.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)


Index(['restaurant_link', 'country', 'price_level', 'avg_rating',
       'total_reviews_count', 'meals_Dinner', 'meals_Lunch', 'meals_Breakfast',
       'meals_Drinks', 'meals_Brunch', 'meals_After-hours',
       'special_diets_Vegetarian Friendly',
       'special_diets_Gluten Free Options', 'special_diets_Vegan Options',
       'special_diets_Halal', 'special_diets_Kosher', 'cuisines_European',
       'cuisines_Italian', 'cuisines_French', 'cuisines_Mediterranean',
       'cuisines_Bar', 'cuisines_Spanish', 'cuisines_Cafe', 'cuisines_Pub',
       'cuisines_Seafood', 'cuisines_Pizza', 'cuisines_British',
       'cuisines_Fast food', 'cuisines_Asian', 'cuisines_Barbecue',
       'features_Seating', 'features_Reservations', 'features_Table Service',
       'features_Wheelchair Accessible', 'features_Serves Alcohol',
       'features_Takeout', 'features_Outdoor Seating',
       'features_Highchairs Available', 'features_Accepts Credit Cards',
       'features_Full Bar', 'features_Parking

In [92]:
# STRONG CORRELATIONS BETWEEN FEATURES

def top_correlations_between_features(country_df, prefixes, threshold):
    features_list = [col for col in country_df.columns if any([pre in col for pre in prefixes])]

    corr_pairs = [(col, country_df[col[0]].corr(country_df[col[1]]))
         for col in itertools.product(*[features_list, features_list])
             if col[0] < col[1] and np.abs(country_df[col[0]].corr(country_df[col[1]])) > threshold
    ]

    corr_pairs = sorted(corr_pairs, key=lambda x: np.abs(x[1]), reverse=True)

    for p in corr_pairs:
        print("{0:.3f}\t{1}".format(p[1], p[0]))
        
    return corr_pairs    

In [93]:
italy_corr_pairs = top_correlations_between_features(
    italy_df2, 
    ['meals_', 'special_diets_', 'cuisines_', 'features_'], 
    0.4
)

0.832	('special_diets_Gluten Free Options', 'special_diets_Vegetarian Friendly')
0.796	('special_diets_Vegan Options', 'special_diets_Vegetarian Friendly')
0.768	('special_diets_Gluten Free Options', 'special_diets_Vegan Options')
0.584	('cuisines_Bar', 'cuisines_Barbecue')
0.566	('features_Free off-street parking', 'features_Parking Available')
0.560	('features_Seating', 'features_Table Service')
0.559	('features_Free Wifi', 'features_Wine and Beer')
0.541	('cuisines_Cafe', 'meals_Breakfast')
0.505	('cuisines_Bar', 'cuisines_Pub')
0.492	('cuisines_Mediterranean', 'cuisines_Seafood')
0.475	('features_Street Parking', 'features_Wine and Beer')
-0.469	('cuisines_Cafe', 'features_Reservations')
0.462	('features_Free off-street parking', 'features_Wine and Beer')
0.439	('features_Accepts Visa', 'features_Free Wifi')
0.437	('features_Reservations', 'meals_Dinner')
-0.433	('cuisines_Cafe', 'meals_Dinner')
0.425	('features_Accepts Visa', 'features_Wine and Beer')
0.418	('features_Reservations

In [112]:
def correlations_between_1_and_set_features(country_df, feature, prefixes, threshold):
    features_list = [col for col in country_df.columns if any([pre in col for pre in prefixes])]

    country_df_threshold = country_df[country_df['total_reviews_count'] > threshold]

    corr_pairs = [(col, country_df_threshold[feature].corr(country_df_threshold[col]))
             for col in features_list
    ]

    corr_pairs = sorted(corr_pairs, key=lambda x: np.abs(x[1]), reverse=True)

    for p in corr_pairs:
        print("{0:.3f}\t{1}".format(p[1], p[0]))
    
    return corr_pairs



In [113]:
italy_avg_rating_corr_pairs = correlations_between_1_and_set_features(
    italy_df2, 
    'avg_rating', 
    ['meals_', 'special_diets_', 'cuisines_', 'features_'], 
    10
)


-0.193	cuisines_Pizza
-0.155	features_Seating
0.154	meals_Brunch
-0.149	meals_After-hours
0.128	meals_Dinner
-0.116	features_Serves Alcohol
0.110	special_diets_Gluten Free Options
0.109	meals_Breakfast
0.102	special_diets_Vegetarian Friendly
0.101	special_diets_Vegan Options
-0.087	features_Table Service
0.086	features_Free Wifi
-0.085	cuisines_Seafood
-0.084	features_Outdoor Seating
0.083	features_Street Parking
0.083	features_Wine and Beer
0.081	meals_Lunch
-0.078	cuisines_Asian
0.055	features_Accepts Visa
0.052	cuisines_Italian
0.049	features_Free off-street parking
0.048	features_Accepts Credit Cards
0.037	cuisines_Bar
0.033	features_Parking Available
-0.028	features_Full Bar
0.026	features_Delivery
-0.025	features_Highchairs Available
-0.023	features_Television
0.022	features_Reservations
0.018	cuisines_French
0.016	cuisines_Barbecue
0.015	cuisines_British
0.014	cuisines_European
0.010	meals_Drinks
0.010	special_diets_Halal
-0.008	cuisines_Spanish
0.008	cuisines_Cafe
0.007	cuisine

**????**

**IDEA**: To divide df into 2 parts (0/1 for some feature) and check correlation between 'avg_rating' and every another feature

<!-- **CONCLUSION**:  -->

<!-- 'avg_rating' doesn't have strong correlation with any feature,  -->

<!-- so we can research the dependencies between 'avg_rating' and other features -->

In [99]:
kk0 = correlations_between_1_and_set_features(
    italy_df2[italy_df2['cuisines_Pizza'] == 0], 
    'avg_rating', 
    ['meals_', 'special_diets_', 'cuisines_', 'features_'], 
    10
)

nan	cuisines_Pizza
-0.152	features_Seating
0.149	meals_Dinner
0.134	meals_Brunch
-0.126	meals_After-hours
0.109	special_diets_Gluten Free Options
-0.109	cuisines_Asian
0.104	special_diets_Vegetarian Friendly
0.099	special_diets_Vegan Options
0.096	cuisines_Italian
-0.091	features_Serves Alcohol
0.085	features_Free Wifi
0.082	features_Wine and Beer
0.081	features_Street Parking
-0.078	features_Outdoor Seating
0.071	meals_Breakfast
0.066	features_Reservations
0.066	meals_Lunch
0.059	features_Accepts Visa
0.054	features_Accepts Credit Cards
0.053	features_Takeout
-0.052	features_Table Service
0.051	features_Delivery
0.050	features_Free off-street parking
0.049	cuisines_Mediterranean
-0.046	cuisines_Seafood
-0.041	features_Full Bar
-0.035	cuisines_Cafe
0.032	features_Parking Available
-0.029	cuisines_Fast food
0.020	cuisines_European
0.016	cuisines_French
-0.015	cuisines_Pub
0.014	cuisines_British
-0.013	cuisines_Spanish
-0.011	meals_Drinks
0.011	cuisines_Bar
0.010	special_diets_Halal
0.01

In [100]:
kk1 = correlations_between_1_and_set_features(
    italy_df2[italy_df2['cuisines_Pizza'] == 1], 
    'avg_rating', 
    ['meals_', 'special_diets_', 'cuisines_', 'features_'], 
    10
)


0.284	meals_Dinner
-0.197	meals_After-hours
0.168	meals_Brunch
nan	special_diets_Halal
nan	special_diets_Kosher
nan	cuisines_Pizza
nan	cuisines_British
-0.217	features_Table Service
-0.209	features_Seating
-0.207	features_Serves Alcohol
-0.130	cuisines_Seafood
0.127	special_diets_Vegan Options
-0.117	features_Outdoor Seating
0.117	special_diets_Gluten Free Options
0.107	special_diets_Vegetarian Friendly
0.105	meals_Breakfast
0.102	meals_Lunch
0.100	features_Street Parking
0.096	features_Wine and Beer
0.087	features_Free Wifi
0.069	cuisines_Fast food
0.062	features_Free off-street parking
-0.057	cuisines_Mediterranean
0.050	features_Accepts Visa
0.049	features_Takeout
0.044	cuisines_Cafe
0.043	features_Delivery
-0.042	features_Highchairs Available
0.041	meals_Drinks
0.035	features_Parking Available
0.031	cuisines_Bar
0.025	features_Wheelchair Accessible
0.025	cuisines_Pub
0.021	cuisines_Italian
0.019	cuisines_Barbecue
0.017	features_Accepts Credit Cards
0.016	cuisines_French
-0.014	feat

In [84]:
it_aggr = pd.DataFrame(italy_df2.groupby(['avg_rating']).size()).sort_index
it_aggr

<bound method DataFrame.sort_index of                 0
avg_rating       
1.0            12
1.5            37
2.0           167
2.5           726
3.0          3067
3.5          9068
4.0         14288
4.5          9352
5.0          1823>

## Bivariative visualizations

In [233]:
import plotly.graph_objects as go

In [290]:
columns_dict = {}
for col_type in ['cuisines', 'meals', 'special_diets', 'features']:
    columns_dict[col_type] = [col for col in italy_df2.columns if col_type+'_' in col and '_cnt' not in col]
    
all_columns = list(set().union(*[val for val in columns_dict.values()]))

df_corr = italy_df2[all_columns].corr(method='pearson')

In [297]:
# cuisines_columns = [col for col in italy_df2.columns if 'cuisines_' in col]
# meals_columns = [col for col in italy_df2.columns if 'meals_' in col]
# special_diets_columns = [col for col in italy_df2.columns if 'special_diets_' in col]
# features_columns = [col for col in italy_df2.columns if 'features_' in col]

# fig = go.Figure(go.Heatmap(
#         z=df_corr.values.tolist(),
#         x=cuisines_columns, #  df_corr.columns
#         y=cuisines_columns,
#         colorscale='rdylgn', 
#         zmin=-1, zmax=1))
# fig.show()

In [296]:
for (var1, var2) in itertools.product(*[columns_dict.keys(), columns_dict.keys()]):
    if var1 <= var2:

        fig = go.Figure(
            data=[go.Heatmap(
                    z=df_corr.loc[columns_dict[var1], columns_dict[var2]].values.tolist(),
                    x=columns_dict[var2], #  df_corr.columns
                    y=columns_dict[var1],
                    colorscale='rdylgn', 
                    zmin=-1, zmax=1,
                )],
            layout=go.Layout(
                title=go.layout.Title(text=f"{var1} vs {var2}")
            )
        )
        fig.show()

<a id="hypothesis"></a>

# 2. Hypothesis Testing

<a id="flags"></a>

# 2.1. For flags

In [107]:
def compare_by_feature(df, feature):
    df2 = df[list(df)]
    df2['summary_score'] = df2['total_reviews_count'] * df2['avg_rating']

    df2_agg = df2.groupby(feature).agg({'restaurant_link': 'count', 'summary_score': 'sum', 'total_reviews_count': 'sum'})
    df2_agg['total_avg_rating'] = df2_agg['summary_score'] / df2_agg['total_reviews_count']

    return df2_agg[['restaurant_link', 'total_avg_rating']]



In [102]:
def filter_with_threshold(df):
#     size of the minimal group
    min_group_size_ratio = df['restaurant_link'].min() / df['restaurant_link'].sum()
#     ratio of difference between ratings (for group with feature and group without feature) and rating of base option (group without feature)
    diff_rating_ratio = abs((df['total_avg_rating'].max() - df['total_avg_rating'].min()) / df['total_avg_rating'].loc[0])
    
    if len(df['total_avg_rating']) == 2:
        if df['total_avg_rating'].loc[0] < df['total_avg_rating'].loc[1]:
            featured_group_rating = 'better'
        elif df['total_avg_rating'].loc[0] > df['total_avg_rating'].loc[1]:
            featured_group_rating = 'worse'
        else:
            featured_group_rating = 'the same'
    else:
        featured_group_rating = 'not applicable'
    return min_group_size_ratio, diff_rating_ratio, featured_group_rating


In [169]:
def print_significant_groups(country_df, feature_prefixes, group_size_threshold, rating_diff_threshold):
    
    features_list = [col for col in country_df.columns if any([pre in col for pre in feature_prefixes]) and '_cnt' not in col]
    
    significance_counter = 0
    
    for feature in features_list:
        df_compare = compare_by_feature(country_df, feature)
        param_filter = filter_with_threshold(df_compare)
        if (param_filter[0] > group_size_threshold 
            and 
            param_filter[1] > rating_diff_threshold
           ):
            significance_counter += 1
            print()
            print()
            print(feature)
            print(df_compare)
            print("Ratio of the smallest group:\t{0:.3f}".format(param_filter[0]))
            print("Variation (??):\t\t\t{0:.3f}\t= {0:.1%}".format(param_filter[1]))
            if param_filter[2] == 'not applicable':
                print(param_filter[2])
            else: 
                print("The featured group is " + str.upper(param_filter[2]))
#             else:
#                 print('the minimal group is too small or the difference < 5%')
            
    return np.sign(significance_counter)

In [232]:
feature_prefixes = ['price_level_', 'meals_', 'special_diets_', 'cuisines_', 'features_']
group_size_threshold = 0.15
rating_diff_threshold = 0.07

for country in list(base_rest_df2['country'].unique()):
    print()
    print()
    print('==========' + '='*len(country) + '==========')
    print('========  ' + str.upper(country) + '  ========')
    print('==========' + '='*len(country) + '==========')
    significance_flg = print_significant_groups(
        features_to_flags(base_rest_df2[base_rest_df2['country'] == country], ['cuisines', 'features'], 0.05), 
        feature_prefixes, group_size_threshold, rating_diff_threshold
    )
    if significance_flg == 0:
        print('There are no significant differences')



========  FRANCE  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)




features_Table Service
                        restaurant_link  total_avg_rating
features_Table Service                                   
0                                  7044          3.617394
1                                 37479          3.898294
Ratio of the smallest group:	0.158
Variation (??):			0.078	= 7.8%
The featured group is BETTER


========  THE NETHERLANDS  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning



features_Free Wifi
                    restaurant_link  total_avg_rating
features_Free Wifi                                   
0                              2956          3.850697
1                               543          4.200117
Ratio of the smallest group:	0.155
Variation (??):			0.091	= 9.1%
The featured group is BETTER


========  GERMANY  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning

There are no significant differences


========  AUSTRIA  ========
There are no significant differences


========  SPAIN  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning



meals_Dinner
              restaurant_link  total_avg_rating
meals_Dinner                                   
0                        5939          3.601546
1                       24692          4.057811
Ratio of the smallest group:	0.194
Variation (??):			0.127	= 12.7%
The featured group is BETTER


meals_Lunch
             restaurant_link  total_avg_rating
meals_Lunch                                   
0                       6918          3.736547
1                      23713          4.042409
Ratio of the smallest group:	0.226
Variation (??):			0.082	= 8.2%
The featured group is BETTER


features_Accepts Credit Cards
                               restaurant_link  total_avg_rating
features_Accepts Credit Cards                                   
0                                        23636          3.771982
1                                         6995          4.145054
Ratio of the smallest group:	0.228
Variation (??):			0.099	= 9.9%
The featured group is BETTER


features_Hi

/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning



special_diets_Vegetarian Friendly
                                   restaurant_link  total_avg_rating
special_diets_Vegetarian Friendly                                   
0                                            21246          3.801449
1                                            14289          4.240235
Ratio of the smallest group:	0.402
Variation (??):			0.115	= 11.5%
The featured group is BETTER


special_diets_Gluten Free Options
                                   restaurant_link  total_avg_rating
special_diets_Gluten Free Options                                   
0                                            23888          3.827800
1                                            11647          4.261214
Ratio of the smallest group:	0.328
Variation (??):			0.113	= 11.3%
The featured group is BETTER


special_diets_Vegan Options
                             restaurant_link  total_avg_rating
special_diets_Vegan Options                                   
0                           

/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)




meals_Dinner
              restaurant_link  total_avg_rating
meals_Dinner                                   
0                       10006          3.572672
1                       28732          4.005614
Ratio of the smallest group:	0.258
Variation (??):			0.121	= 12.1%
The featured group is BETTER


meals_Lunch
             restaurant_link  total_avg_rating
meals_Lunch                                   
0                      13074          3.625342
1                      25664          4.006944
Ratio of the smallest group:	0.337
Variation (??):			0.105	= 10.5%
The featured group is BETTER


cuisines_Pizza
                restaurant_link  total_avg_rating
cuisines_Pizza                                   
0                         30215          3.981320
1                          8523          3.695881
Ratio of the smallest group:	0.220
Variation (??):			0.072	= 7.2%
The featured group is WORSE


========  IRELAND  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning



special_diets_Vegetarian Friendly
                                   restaurant_link  total_avg_rating
special_diets_Vegetarian Friendly                                   
0                                             1316          4.007908
1                                              882          4.358300
Ratio of the smallest group:	0.401
Variation (??):			0.087	= 8.7%
The featured group is BETTER


special_diets_Gluten Free Options
                                   restaurant_link  total_avg_rating
special_diets_Gluten Free Options                                   
0                                             1426          3.977732
1                                              772          4.372405
Ratio of the smallest group:	0.351
Variation (??):			0.099	= 9.9%
The featured group is BETTER


special_diets_Vegan Options
                             restaurant_link  total_avg_rating
special_diets_Vegan Options                                   
0                             

/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning



cuisines_Scandinavian
                       restaurant_link  total_avg_rating
cuisines_Scandinavian                                   
0                                 1047          3.777093
1                                  283          4.054728
Ratio of the smallest group:	0.213
Variation (??):			0.074	= 7.4%
The featured group is BETTER


features_Accepts Credit Cards
                               restaurant_link  total_avg_rating
features_Accepts Credit Cards                                   
0                                         1003          3.644505
1                                          327          4.120391
Ratio of the smallest group:	0.246
Variation (??):			0.131	= 13.1%
The featured group is BETTER


========  BELGIUM  ========
There are no significant differences


========  GREECE  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning



special_diets_Vegetarian Friendly
                                   restaurant_link  total_avg_rating
special_diets_Vegetarian Friendly                                   
0                                             3550          4.097346
1                                             1523          4.419690
Ratio of the smallest group:	0.300
Variation (??):			0.079	= 7.9%
The featured group is BETTER


special_diets_Gluten Free Options
                                   restaurant_link  total_avg_rating
special_diets_Gluten Free Options                                   
0                                             4037          4.135162
1                                             1036          4.451087
Ratio of the smallest group:	0.204
Variation (??):			0.076	= 7.6%
The featured group is BETTER


special_diets_Vegan Options
                             restaurant_link  total_avg_rating
special_diets_Vegan Options                                   
0                             

/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning



features_Free Wifi
                    restaurant_link  total_avg_rating
features_Free Wifi                                   
0                              6109          3.961888
1                              2209          4.243567
Ratio of the smallest group:	0.266
Variation (??):			0.071	= 7.1%
The featured group is BETTER


========  SLOVAKIA  ========


cuisines_Cafe
               restaurant_link  total_avg_rating
cuisines_Cafe                                   
0                          418          3.948017
1                           99          4.236649
Ratio of the smallest group:	0.191
Variation (??):			0.073	= 7.3%
The featured group is BETTER


features_Free Wifi
                    restaurant_link  total_avg_rating
features_Free Wifi                                   
0                               313          3.836259
1                               204          4.113887
Ratio of the smallest group:	0.395
Variation (??):			0.072	= 7.2%
The featured group is BETTE

/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning



features_Takeout
                  restaurant_link  total_avg_rating
features_Takeout                                   
0                            1383          3.965098
1                             631          4.276042
Ratio of the smallest group:	0.313
Variation (??):			0.078	= 7.8%
The featured group is BETTER


features_Free Wifi
                    restaurant_link  total_avg_rating
features_Free Wifi                                   
0                              1387          3.788282
1                               627          4.233938
Ratio of the smallest group:	0.311
Variation (??):			0.118	= 11.8%
The featured group is BETTER


features_Full Bar
                   restaurant_link  total_avg_rating
features_Full Bar                                   
0                             1424          3.858919
1                              590          4.143604
Ratio of the smallest group:	0.293
Variation (??):			0.074	= 7.4%
The featured group is BETTER


features_Accepts

/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning

There are no significant differences


========  DENMARK  ========


features_Accepts Credit Cards
                               restaurant_link  total_avg_rating
features_Accepts Credit Cards                                   
0                                          929          3.849343
1                                          335          4.137330
Ratio of the smallest group:	0.265
Variation (??):			0.075	= 7.5%
The featured group is BETTER


========  BULGARIA  ========


meals_Lunch
             restaurant_link  total_avg_rating
meals_Lunch                                   
0                        104          3.858023
1                        432          4.218663
Ratio of the smallest group:	0.194
Variation (??):			0.093	= 9.3%
The featured group is BETTER


meals_Brunch
              restaurant_link  total_avg_rating
meals_Brunch                                   
0                         389          4.003463
1                         147          4.325418
Ratio of th

/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning

**IDEA**: TO CHECK CORRELATION BETWEEN SIGNIFICANT FEATURES

kek:

==================================
========  CZECH REPUBLIC  ========
==================================
cuisines_Czech
                restaurant_link  total_avg_rating
cuisines_Czech                                   
0                           807          4.255940
1                          1207          3.971821
Ratio of the smallest group:	0.401
Variation (??):			0.067
The featured group is WORSE


===========================
========  ROMANIA  ========
===========================

cuisines_Romanian
                   restaurant_link  total_avg_rating
cuisines_Romanian                                   
0                              624          4.206314
1                              499          3.879610
Ratio of the smallest group:	0.444
Variation (??):			0.078
The featured group is WORSE


In [ ]:
# A value is trying to be set on a copy of a slice from a DataFrame.
# Try using .loc[row_indexer,col_indexer] = value instead

# 1)
# data[data.bidder == 'parakeet2004']['bidderrate'] = 100
# -> 
# data.loc[data.bidder == 'parakeet2004', 'bidderrate'] = 100

# 2)
# data['bidtime_hours'] = data.bidtime.map(lambda x: x * 24)
# ->
# data.loc[:, 'bidtime_hours'] = data.bidtime.apply(lambda x: x * 24)


<a id="flags"></a>

# 2.2. For 3 types by cuisines (only/one of/no)

In [214]:
def define_feature_type(df, feature):
    feature_category = feature[:feature.rfind('_')]
    if df[feature] == 0:
        return 'no'
    elif df[feature] == 1 and df[feature_category + '_cnt'] == 1:
        return 'only'
    else:
        return 'one of'

def define_types_by_cuisine(df, feature): 
    df['feature_type'] = df.apply(lambda x: define_feature_type(x, feature), axis=1)
    return df

In [215]:
def filter_with_threshold2(df):
    
    if all(option in df.index for option in ['only', 'no']):
#         size of the minimal group
        min_group_size_ratio = df.loc['only', 'restaurant_link'] / df['restaurant_link'].sum()

#         ratio of difference between ratings (for group with feature and group without feature) and rating of base option (group without feature)
        diff_rating_ratio = abs((df.loc['only', 'total_avg_rating'] - df.loc['no', 'total_avg_rating']) / df.loc['only', 'total_avg_rating'])
        
        if df.loc['no', 'total_avg_rating'] < df.loc['only', 'total_avg_rating']:
            featured_group_rating = 'better'
        elif df.loc['no', 'total_avg_rating'] > df.loc['only', 'total_avg_rating']:
            featured_group_rating = 'worse'
        else:
            featured_group_rating = 'the same'
    else:
        min_group_size_ratio = 0
        diff_rating_ratio = 0
        featured_group_rating = 'not applicable'
    return min_group_size_ratio, diff_rating_ratio, featured_group_rating


In [224]:
def print_significant_groups2(country_df, feature_prefixes, group_size_threshold, rating_diff_threshold):
    
    features_list = [col for col in country_df.columns if any([pre in col for pre in feature_prefixes]) and '_cnt' not in col]
    significance_counter = 0
    
    for feature in features_list:
        country_df2 = define_types_by_cuisine(country_df, feature) 
        df_compare = compare_by_feature(country_df2, 'feature_type')
        param_filter = filter_with_threshold2(df_compare)
        if (param_filter[0] > group_size_threshold 
            and 
            param_filter[1] > rating_diff_threshold
           ):
            significance_counter += 1
            print()
            print()
            print(feature)
            print(df_compare)
            print("Ratio of the smallest group:\t{0:.3f}".format(param_filter[0]))
            print("Variation (??):\t\t\t{0:.3f}\t= {0:.1%}".format(param_filter[1]))
            if param_filter[2] == 'not applicable':
                print(param_filter[2])
            else:
                print("The featured group is " + str.upper(param_filter[2]))
            
    return significance_counter

In [231]:
feature_prefixes = ['price_level_', 'meals_', 'special_diets_', 'cuisines_']
group_size_threshold = 0.05
rating_diff_threshold = 0.05

for country in list(base_rest_df2['country'].unique()):
    print()
    print()
    print('==========' + '='*len(country) + '==========')
    print('========  ' + str.upper(country) + '  ========')
    print('==========' + '='*len(country) + '==========')
    significance_flg = print_significant_groups2(
        features_to_flags(base_rest_df2[base_rest_df2['country'] == country], ['cuisines', 'features'], 0.05), 
        feature_prefixes
        , group_size_threshold, rating_diff_threshold
    )
    if significance_flg == 0:
        print('There are no significant differences')



========  FRANCE  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)




cuisines_French
              restaurant_link  total_avg_rating
feature_type                                   
no                      12464          3.682730
one of                  15981          3.914534
only                    16078          3.932702
Ratio of the smallest group:	0.361
Variation (??):			0.064	= 6.4%
The featured group is BETTER


========  THE NETHERLANDS  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning

There are no significant differences


========  GERMANY  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning

There are no significant differences


========  AUSTRIA  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning

There are no significant differences


========  SPAIN  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)




cuisines_Spanish
              restaurant_link  total_avg_rating
feature_type                                   
no                       7823          4.123258
one of                  15505          4.002390
only                     7303          3.716782
Ratio of the smallest group:	0.238
Variation (??):			0.109	= 10.9%
The featured group is WORSE


========  UNITED KINGDOM  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning



cuisines_Cafe
              restaurant_link  total_avg_rating
feature_type                                   
no                      27640          4.166338
one of                   4882          4.214355
only                     3013          3.842749
Ratio of the smallest group:	0.085
Variation (??):			0.084	= 8.4%
The featured group is WORSE


price_level_Cheap Eats
              restaurant_link  total_avg_rating
feature_type                                   
no                      23002          4.197097
only                    12533          3.909030
Ratio of the smallest group:	0.353
Variation (??):			0.074	= 7.4%
The featured group is WORSE


========  ITALY  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)


There are no significant differences


========  IRELAND  ========


cuisines_Cafe
              restaurant_link  total_avg_rating
feature_type                                   
no                       1767          4.300455
one of                    248          4.307373
only                      183          4.006571
Ratio of the smallest group:	0.083
Variation (??):			0.073	= 7.3%
The featured group is WORSE


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)




========  POLAND  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning

There are no significant differences


========  SWEDEN  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning

There are no significant differences


========  BELGIUM  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning

There are no significant differences


========  GREECE  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning

There are no significant differences


========  CROATIA  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning

There are no significant differences


========  PORTUGAL  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)


There are no significant differences


========  SLOVAKIA  ========


cuisines_Cafe
              restaurant_link  total_avg_rating
feature_type                                   
no                        418          3.948017
one of                     51          4.194201
only                       48          4.315402
Ratio of the smallest group:	0.093
Variation (??):			0.085	= 8.5%
The featured group is BETTER


========  CZECH REPUBLIC  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning



cuisines_Cafe
              restaurant_link  total_avg_rating
feature_type                                   
no                       1711          4.063769
one of                    186          4.205226
only                      117          3.697278
Ratio of the smallest group:	0.058
Variation (??):			0.099	= 9.9%
The featured group is WORSE


price_level_Cheap Eats
              restaurant_link  total_avg_rating
feature_type                                   
no                       1595          4.037758
only                      419          4.258728
Ratio of the smallest group:	0.208
Variation (??):			0.052	= 5.2%
The featured group is BETTER


========  HUNGARY  ========


cuisines_Cafe
              restaurant_link  total_avg_rating
feature_type                                   
no                        721          4.261764
one of                    115          4.098114
only                       49          3.861927
Ratio of the smallest group:	0.055
Variation (??):		

/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning



========  FINLAND  ========


cuisines_Cafe
              restaurant_link  total_avg_rating
feature_type                                   
no                        528          3.941487
one of                     78          4.092890
only                       42          4.155119
Ratio of the smallest group:	0.065
Variation (??):			0.051	= 5.1%
The featured group is BETTER


========  DENMARK  ========


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning

There are no significant differences


========  BULGARIA  ========
There are no significant differences


========  ROMANIA  ========
There are no significant differences


/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, col + '_' + t] = df[col].apply(lambda x: 1 if t in str(x) else 0)
/var/folders/zh/mnvg84rj3_7413ycjzpn5y8r0000gp/T/ipykernel_75787/128475807.py:22: SettingWithCopyWarning

=========================
========  SPAIN  ========
=========================

cuisines_Spanish
              restaurant_link  total_avg_rating
cuisine_type                                   
no                       7823          4.123258
one of                  15505          4.002390
only                     7303          3.716782
Ratio of the smallest group:	0.238
Variation (??):			0.109	= 10.9%
The featured group is WORSE


==================================
========  UNITED KINGDOM  ========
==================================

price_level_Cheap Eats
              restaurant_link  total_avg_rating
feature_type                                   
no                      23002          4.197097
only                    12533          3.909030
Ratio of the smallest group:	0.353
Variation (??):			0.074	= 7.4%
The featured group is WORSE

<a id="conclusions"></a>

# 3. Conclusions

# =========================== OLD:

In [ ]:
'''
IDEALLY: if I compare by 1 feature, other features must be the same

BUT: restaurants with 'special_diets' are BETTER almost always
'''

In [52]:
# import seaborn as sns
# import matplotlib.pyplot as plt

In [72]:
# sns.histplot(data=bavaria_df[bavaria_df['total_reviews_count'] >= 20], x='total_reviews_count', bins=200)
# plt.xlim([20, 2000])
# plt.ylim([0, 1000])
# plt.show()